### PR TITLE
[Merged by Bors] - Add support for LDAP authentication

### PR DESCRIPTION
**Pull request by sbernauer**
## Description

Superseding https://github.com/stackabletech/superset-operator/pull/145, completely reworked to use new configuration mechanism introduced in https://github.com/stackabletech/superset-operator/pull/173
For https://github.com/stackabletech/superset-operator/issues/5
Integration-Tests: https://github.com/stackabletech/integration-tests/pull/175
Use `./create_test_cluster.py --kind --operator superset=0.4.0-pr180` before running the tests

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by sbernauer** [comment]
Integration tests are green => ready for review :)

File: rust/operator-binary/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        "os.environ.get('SQLALCHEMY_DATABASE_URI')".into(),
+    );
+    config.insert(
+        SupersetConfigOptions::StatsLogger.to_string(),
+        "StatsdStatsLogger(host='0.0.0.0', port=9125)".into(),
+    );
+    config.insert(
+        SupersetConfigOptions::MapboxApiKey.to_string(),
+        "os.environ.get('MAPBOX_API_KEY')".into(),
+    );
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/config.rs:39]: Smoke tests do all pass not it seems to work fine. Solution is far from ideal anyway.

**Review by maltesander — COMMENTED** [review]
Just a small thing. Ran the tests and works. Maybe @siegfriedweber can have another look? Im not that into ldap and superset ;)

@sbernauer i cannot judge how good i is, but since we will required ldap etc in other operators to i think its fine to merge in a working approach and refine / improve that later as well.

File: rust/operator-binary/src/superset_controller.rs [hunk 2]
```diff
@@ -476,13 +517,88 @@ fn build_server_rolegroup_statefulset(
...
+        AuthenticationClassProvider::Ldap(ldap) => {
+            if let Some(bind_credentials) = &ldap.bind_credentials {
+                let volume_name = format!("{authentication_class_name}-bind-credentials");
+
+                pb.add_volume(build_secret_operator_volume(
+                    &volume_name,
+                    &bind_credentials.secret_class,
+                    bind_credentials.scope.as_ref(),
+                ));
+                cb.add_volume_mount(&volume_name, format!("/stackable/secrets/{volume_name}"));
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/superset_controller.rs]: Good point, thx

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/config.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/config.rs:40]: Good hint, thanks!

File: examples/superset-with-ldap.yaml [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      containers:
+        - name: openldap
+          image: docker.io/bitnami/openldap:2.5
+          env:
+            - name: LDAP_ADMIN_USERNAME
+              value: admin
+            - name: LDAP_ADMIN_PASSWORD
+              value: admin
+            - name: LDAP_USERS
+              value: user01,user02
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply examples/superset-with-ldap.yaml:47]: Authorization was intentionally not included in this PR. I know that having all user assigned the Public role is not really usable but it's a secure default at the moment. The role can be set by the user to e.g. make all users Gamma.
I would expect Authorization (e.g. via LDAP group memberships) in a follow-up PR.
The line you commented here is btw from the bitnami opernldap container and is just to create 2 testusers in the LDAP instance.

File: examples/superset-with-ldap.yaml [hunk 4]
```diff
@@ -0,0 +1,160 @@
+# helm install --repo https://repo.stackable.tech/repository/helm-stable/ secret-operator secret-operator
+# helm install --repo https://repo.stackable.tech/repository/helm-stable/ commons-operator commons-operator
+# Until commons-operator is available officially: helm install --version 0.1.0-pr11 --repo https://repo.stackable.tech/repository/helm-test/ commons-operator commons-operator
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply examples/superset-with-ldap.yaml:3]: ack

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -94,9 +132,67 @@ pub struct SupersetClusterSpec {
...
+    /// See [Flask LDAP documentation](https://flask-appbuilder.readthedocs.io/en/latest/security.html#authentication-ldap)
+    pub ldap_extras: Option<SupersetClusterAuthenticationConfigLdapExtras>,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+/// Additional configs to LDAP Authentication in Superset. See <https://flask-appbuilder.readthedocs.io/en/latest/security.html#authentication-ldap>
+pub struct SupersetClusterAuthenticationConfigLdapExtras {
+    /// Allow users who are not already in the FAB DB.
+    /// Gets mapped to `AUTH_USER_REGISTRATION`
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:162]: Test comment

File: rust/crd/src/lib.rs [hunk 6]
```diff
@@ -94,9 +132,68 @@ pub struct SupersetClusterSpec {
...
     #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub authentication_config: Option<SupersetClusterAuthenticationConfig>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
     pub nodes: Option<Role<SupersetConfig>>,
 }
 
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SupersetClusterAuthenticationConfig {
+    pub methods: Vec<SupersetClusterAuthenticationConfigMethod>,
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/crd/src/lib.rs:143]: As other operators also don't have an simple list (e.g. current Zookeeper discussion) we will make this an `authenticationConfig: Option<SupersetClusterAuthenticationConfigMethod>`

[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/crd/src/lib.rs:143]: We agreeed on 
```
    authenticationConfig: # optional
      authenticationClass: superset-with-ldap-server-veri-tls-ldap # optional. If not set, AUTH_DB is used
      userRegistration: true # default true
      userRegistrationRole: Public # default Public

File: examples/superset-with-ldap.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply examples/superset-with-ldap.yaml:47]: We add the Admin role in the example

File: examples/superset-with-ldap.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply examples/superset-with-ldap.yaml:3]: commons-operator 0.1.0 released :)

File: examples/superset-with-ldap.yaml [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+---
+apiVersion: secrets.stackable.tech/v1alpha1
+kind: SecretClass
+metadata:
+  name: openldap-tls
+spec:
+  backend:
+    autoTls:
+      ca:
+        # autoGenerate: true # Needed in a future secret-operator version
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply examples/superset-with-ldap.yaml]: secret-operator 0.3.0 released, example adopted

File: rust/crd/src/lib.rs [hunk 8]
```diff
@@ -94,9 +132,68 @@ pub struct SupersetClusterSpec {
...
+
+    /// This role will be given in addition to any AUTH_ROLES_MAPPING.
+    /// Gets mapped to `AUTH_USER_REGISTRATION_ROLE`
+    #[serde(default = "default_user_registration_role")]
+    pub user_registration_role: String,
+
+    /// If we should replace ALL the user's roles each login, or only on registration.
+    /// Gets mapped to `AUTH_ROLES_SYNC_AT_LOGIN`
+    #[serde(default = "default_sync_roles_at")]
+    pub sync_roles_at: LdapRolesSyncMoment,
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/crd/src/lib.rs:161]: With the default value they are effective optional with the benefit of having well-documented defaults in the CRD.

**Comment by sbernauer** [comment]
@siegfriedweber could you please take another look?

**Review by siegfriedweber — CHANGES_REQUESTED** [review]
There are some minor and one major issue left.

The major issue is that some views are not accessible by the administrator, e.g. the user list. The reason is that the `Admin` role does not have the `UserLDAPModelView` permission. The command `superset init` which is called in the job created by the `superset_db_controller` initializes the roles according to the configured authentication type. If the authentication type is `AUTH_LDAP` then the permission `UserLDAPModelView` is added to the `Admin` role. Unfortunately the authentication type is not set in the configuration created by the `superset_db_controller` and it is not trivial to add it.

**Comment by sbernauer** [comment]
> The major issue is that some views are not accessible by the administrator, e.g. the user list. 

The problem should now be sorted out via https://github.com/stackabletech/superset-operator/pull/180/commits/458b7f2e0f4520ea98b37c3849769ad6cb138ed1

The problem with the first login failing is purely a bug in flask-appbuilder and should be fixed by https://github.com/dpgaspar/Flask-AppBuilder/pull/1846 which is entirely a docker image thing and not part of this PR.

Could you please take another look - would be great before your holiday to get this in :innocent:

File: rust/operator-binary/src/main.rs [hunk 9]
```diff
@@ -229,3 +251,17 @@ async fn main() -> anyhow::Result<()> {
...
+    authentication_class: &AuthenticationClass,
+) -> bool {
+    if let Some(authentication_config) = authentication_config {
+        if let Some(authentication_class_name) = &authentication_config.authentication_class {
+            return authentication_class_name
+                == authentication_class.metadata.name.as_ref().unwrap();
+        }
+    }
+
+    false
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/operator-binary/src/main.rs]: Thanks for the tip!

**Review by adwk67 — APPROVED** [review]
lgtm. Opened a separate issue #195 to avoid db-clashes in the tests

**Comment by sbernauer** [comment]
Thanks all for your feedback!
Everything should work to the best of my knowledge. Together with https://github.com/stackabletech/docker-images/pull/116 this seems to be a nice solution :)
The Authorization part is outstanding and will be handled via a separate PR.

**Pull request closed without merging**